062-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,176 @@
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. This prevents using metrics like an API, whether this is to build grafana dashboards or for the operator to rely on metrics.
+1. Due to the complex rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions.
+
+
+## Proposal
+
+The proposal is to build metrics reporters that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/052-proposal.png)
+
+Today to enable jmx_exporter, users use:
+```
+metricsConfig:
+  type: jmxPrometheusExporter
+  valueFrom:
+    configMapKeyRef:
+      name: kafka-metrics
+      key: kafka-metrics-config.yml
+```
+
+To enable metric reporters, they will instead use this for example:
+```
+metricsConfig:
+  type: strimziMetricsReporter
+  config:
+    allowList: ".*"
+```
+
+Reporters will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
+
+Strimzi should set the port to 9404 to make the migration from jmx_exporter easy.
+
+The reporter will also export JVM metrics similar to the ones exported by jmx_exporter. These are provided by the [Hotspot exports](io.prometheus.client.hotspot) from the Prometheus Java client.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it now or later to the reporter.
+
+This proposal will produce an implementation for each type of metrics reporter.
+- `KafkaPrometheusMetricsReporter` usable on brokers (for Kafka metrics) and on clients (including Connect and Streams)
+- `YammerPrometheusMetricsReporter` usable on brokers (for Yammer metrics)
+
+The Prometheus metrics registry is a singleton and the HTTP server will also be a singleton. This will allow applications to start multiple instances of the reporter (for example in applications containing multiple Kafka clients like Streams, Connect, Strimzi HTTP bridge, or in application loading plugins like the OAuth plugin that instance their own reporter instances), and still collect all metrics via a single HTTP endpoint per JVM. 
```

Comment:
You mention the Bridge -> but using this in the Bridge would require to be also able to plug into the reporter applications that are completely unrelated to Kafka itself such as some HTTP server or application. Is that the plan? Can you clarify that?

Reply:
I've added a section detailing the changes required in the Bridge. I hope this addresses your concerns.